### PR TITLE
[vLLM] Require explicit decode reload commands

### PR DESCRIPTION
**Pull request by tchedaTT**
## Summary

- implement the strict version-1 decode update contract: every refactored call supplies `reload_inputs`, `reload_page_table`, `reload_sampling_params`, and `reset_sampling_state`
- advertise `decode_input_update_contract = 1` on production adapters consumed by the standalone plugin
- remove model-side reload heuristics and move standalone demo and SGLang policy to explicit call sites
- preserve Gemma4 batch-keyed traces, hybrid page-table routing, and padded token feedback while making mode and layout reloads plugin-owned
- apply each `slot_remap` exactly once to persistent slot-indexed state, including dormant sampling state; split lane-DP mappings into rank-local permutations
- preserve continuing DeepSeek parameters, penalty history, and seed state during partial prefill
- keep Galaxy parameter shadows aligned with slot remaps and align seed counters only from authoritative host positions
- avoid warmup penalty-state resets when no request-owned history exists
- initialize and upload first-lifecycle seeds directly, including `seed=None`, without relying on tenstorrent/tt-metal#51556

## Negotiation and rollout

The paired standalone-plugin change, tenstorrent/vllm-tt-plugin#78, is merged. The plugin preserves the legacy `reset_batch` call shape, with a warning, for unmarked version-0 adapters. Marked adapters receive the explicit version-1 commands from this PR. Refactored model APIs remain strict: direct callers provide all four commands and no model-side fallback heuristics are restored.

`model_capabilities["supports_async_decode"]` is a separate residency gate. A wrapper may enable it only when sampled-token feedback remains device-resident and persistent position advances exactly once. Other wrappers receive explicit full-input reload commands from the plugin.

## Design specification

- [tt-metal sampling and decode contract](https://github.com/tenstorrent/tt-metal/blob/tcheda/issue-449-reload-contract/models/common/sampling/README.md)
- [merged plugin reload-contract specification](https://github.com/tenstorrent/vllm-tt-plugin/blob/main/docs/DECODE_RELOAD_CONTRACT.md)

These documents define the command semantics, transition requirements, remap and sampling-state rules, warmup and SGLang policies, and the model-author checklist.

## Rebase and review fixes

Rebased onto tt-metal `main` at `68107a0c9b14f3922c53371f0e69a877168ac2d6`; PR head is `8b263d2239f5283c9ddef991032235afbe729c5b`.

- the removed T3000 vLLM adapter remains deleted
- mode changes and decode layout or batch changes cause a plugin-planned full reload before selecting another trace buffer
- Galaxy separated sampling now receives the forward input-authority command; seed counters realign only when host positions are authoritative
- DeepSeek partial admission preserves unlisted live slots and resets penalty state only for incoming slots
- Galaxy slot remaps move the per-slot parameter shadow and sampler state together

## Validation

- repository pre-commit hooks passed locally over the complete PR diff after the final rebase
- Python syntax compilation and `git diff --check` passed after the final rebase
- focused regressions cover strict direct callers, mode and Gemma4 batch trace selection, DeepSeek partial-admission state preservation, lane-DP remapping, Galaxy seed and parameter remapping, reset-only seeds, warmup behavior, and Gemma4 PLI token feedback
- current PR CI: [checks for this branch](https://github.com/tenstorrent/tt-metal/actions?query=branch%3Atcheda%2Fissue-449-reload-contract)
- refreshed all-model, all-SKU, all-tier vLLM branch nightly against merged plugin main `7250ddfaa988cc7417f518266dce52e425745472`: [run 35680811729](https://github.com/tenstorrent/tt-metal/actions/runs/35680811729) (running)

Related to tenstorrent/vllm#449

### CI Status

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:tcheda/issue-449-reload-contract)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:tcheda/issue-449-reload-contract)

### CODE REVIEW

**Comment by tchedaTT** [comment]
Branch nightly for this PR:

- https://github.com/tenstorrent/tt-metal/actions/runs/33196652335

Refs: tt-metal 3c9892f14b4aa9e6c1e8c0138cefcf76d6d4dd8b, vllm-tt-plugin#78 3a7e9646f5cda42c65c1dffa83cc7a4f80603600. All configured models, SKUs, and tiers.

**Review by arminaleTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -1585,110 +1576,33 @@ def decode_forward(
...
-        reset_batch=False,
         prompt_tokens: torch.Tensor | None = None,
         output_tokens: torch.Tensor | None = None,
         slot_remap=None,
         defer_device_sampling: bool = False,
+        *,
+        reload_inputs: bool,
+        reload_page_table: bool,
+        reload_sampling_params: bool,
+        reset_sampling_state: bool,
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/tt_transformers/tt/generator.py:1799]: Fixed in 6279c623aee. The shared signature remains strict. I migrated the GRPO worker, both Quasar GPT-OSS demo paths, its accuracy test, and the Quasar Qwen3-VL demo to send all four commands explicitly. The GRPO worker keeps its old every-step parameter upload and only resets history on its first decode. An AST regression now covers every migrated direct caller.

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -2153,14 +2089,12 @@ def _decode_token_feedback_buffer(model, device_inputs):
...
-        not be a valid sampling output (gemma4's is rank-2; ``ttnn.sampling``
-        requires a rank-4 preallocated output). Returning None makes sampling
-        allocate its own output instead of writing into ``device_inputs[0]``.
+        Some models' token input buffer is not a valid sampling output
+        (gemma4's is rank-2; ``ttnn.sampling`` requires a rank-4 preallocated
+        output). Returning None makes sampling allocate its own output instead
+        of writing into ``device_inputs[0]``.
         """
-        if getattr(model, "_tt_vllm_always_refresh_decode_trace_inputs", False):
+        if not getattr(model, "_tt_supports_decode_token_feedback", True):
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/tt_transformers/tt/generator.py:2340]: Fixed in 6279c623aee. Gemma4 now has an explicit fail-closed class capability and enables decode token feedback only for non-PLI variants whose token input has the rank-4 sampling-output layout. The runtime kill switch also disables feedback. I added a contract regression; the equivalent Quasar Qwen3-VL opt-out is explicit as well.

File: models/demos/deepseek_v3/tt/generator.py [hunk 3]
```diff
@@ -947,23 +962,65 @@ def _tt_from_tokens_step(self, tokens_step: torch.Tensor) -> ttnn.Tensor:
...
+            output_tokens=output_tokens_device,
         )
+        seed_manager = self.sampling_generator.seed_manager
+        active_seed_slots = self._sampling_device_slots(user_slots)
+        if active_seed_slots is None:
+            active_seed_slots = list(range(seed_manager.max_batch_size))
+        # A request finishing at the batch tail produces no non-identity
+        # remap, so retire seed/salt state that no longer belongs to a live
+        # row before assigning salts to newly admitted requests.
+        seed_manager.deactivate_slots_except(active_seed_slots)
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/demos/deepseek_v3/tt/generator.py]: Fixed in 6279c623aee. DeepSeek partial prefill now explicitly sets preserve_unlisted_slots=True. Incoming stable slots are still reset and reseeded, but they are no longer passed to deactivate_slots_except as though they were the complete live set. The regression verifies the row-padded incoming slots and that no continuing slot is deactivated.

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 4]
```diff
@@ -1659,61 +1608,74 @@ def _remember_slot_params(self, params):
...
 
+    def _apply_sampling_slot_remap(self, slot_remap) -> None:
+        if slot_remap is None:
+            return
+        sampling_module = getattr(self.model, "sampling", None)
+        if sampling_module is None:
+            return
+        seed_manager = sampling_module.seed_manager
+        sm_bs = seed_manager.max_batch_size
+        seed_manager.apply_slot_remap(slot_remap[0:sm_bs])
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Fixed in 6279c623aee. Galaxy now snapshots and remaps every cached per-slot parameter vector with the same destination-to-source permutation as SeedManager. The shadow is updated only after remap validation and the seed remap succeeds. A swap/condense-style regression verifies temperature and top-k move with seed state.

File: models/demos/deepseek_v3/tt/generator.py [hunk 5]
```diff
@@ -947,23 +962,65 @@ def _tt_from_tokens_step(self, tokens_step: torch.Tensor) -> ttnn.Tensor:
...
+        output_tokens_device = self._sampling_device_history(output_tokens) if output_tokens is not None else None
         self.sampling_generator.apply_decode_state(
             sampling_param_chunks,
-            reset_batch=True,
-            prompt_tokens=torch.zeros((batch_size_per_row, 1), dtype=torch.int64),
-            output_tokens=torch.zeros((batch_size_per_row, 1), dtype=torch.int64),
+            reload_sampling_params=reload_sampling_params,
+            reset_sampling_state=reset_sampling_state,
+            prompt_tokens=prompt_tokens_device,
+            output_tokens=output_tokens_device,
```
[the remark this answers is not in the document]
Comment (tchedaTT) [reply models/demos/deepseek_v3/tt/generator.py:1011]: Fixed in 6279c623aee. Partial DeepSeek prefill passes its row-padded incoming slots through SamplingGenerator.apply_decode_state. Prompt history is updated only for those slots, and output penalty counts are cleared with a broadcast device row mask that leaves continuing rows unchanged. Full authoritative decode resets still rebuild the whole buffer. Regressions cover slot-scoped routing and selective output clearing.

**Comment by tchedaTT** [comment]
Fresh branch nightly after the five review fixes:

- [vLLM Model Tests — all models, SKUs, and tiers](https://github.com/tenstorrent/tt-metal/actions/runs/33198503605)
- tt-metal: 6279c623aee571835e54f2b6901a2230ea92d21f
- vllm-tt-plugin PR #78: 3a7e9646f5cda42c65c1dffa83cc7a4f80603600

**Review by yieldthought — APPROVED** [review]

**Review by bbradelTT — APPROVED** [review]

File: models/demos/t3000/llama2_70b/tt/generator_vllm.py
Comment (mtairum) [inline models/demos/t3000/llama2_70b/tt/generator_vllm.py:1]: This file was removed yesterday. Please rebase to latest main.

**Comment by mtairum** [comment]
Requesting a rebase and a fresh CI run before this lands.

**1. Branch conflicts with main.** `mergeable` is now `CONFLICTING` — one modify/delete conflict: `models/demos/t3000/llama2_70b/tt/generator_vllm.py` was removed on main by `70aaacb6cd7` (#54452) and is modified here. Main has moved a good way past the merge base (`69e7e920fc9`) too.

**2. Please re-run the vLLM/model CI at the rebased head and link it in the description.** The description links runs at `6279c623aee` (08-28). Head is `f3f318f97b7`, two commits later — `57ae18de6a8` and `f3f318f97b7`, both in the shared `SamplingGenerator`. There *are* runs at head (`33620073299` and five others) but they aren't linked, and they show a regression against the validated SHA.

**Main blocker: `vllm-tests / Gemma4-31B [bh_quietbox_2]` regressed at head.**

Same job, same SKU, present in all three matrices:

- main `1b1ee1c9a3c` (08-28, run `33196649113`) — success
- `6279c623aee` (08-28, run `33198503605`) — success  ← the SHA the description validates
- `f3f318f97b7` (09-02, run `33620073299`) — **failure**

It's a hang, not a crash: `❌ Server did not become ready in time (1200s)`, no fatal marker, launcher still alive. The server log goes quiet at 11:30:51 after the weight cache builds at 11:27:52, with `enable_model_warmup=True`, `sample_on_device_mode=decode_only`, `trace_mode=all`. Gemma4-12B, 26B-A4B, E2B and 31B on `wh_llmbox` all still pass, so it looks specific to 31B on Blackhole. The only deltas between the green SHA and the red one are the main merge plus the two sampling commits, so a rebase and re-run should isolate it.

**Likely mechanism worth checking while you're in there.** Three per-model reload guards were removed with no replacement, and each covers a decision the plugin can't make, because it depends on trace-key granularity only the model side can see:

- `sampling_mode_changed` — `trace_inputs_decode` is keyed by `on_device_sampling` (`tt_transformers/tt/generator.py:2110`), so each mode owns separate device token/position/page-table buffers. Pre-PR `reload_inputs` included `or sampling_mode_changed` (`main:1823-1832`). A host<->device sampling flip with unchanged layout now gives the plugin no reason to set `reload_inputs=True`, and the trace replays the other mode's buffers.
- `batch_changed` (gemma4) — traces keyed by `(mode, batch)`. Pre-PR: `reset_inputs = reset_batch or not on_device_sampling or sampling_mode_changed or batch_changed` (`main:1461-1462`). `_prev_decode_batch` is still written (`gemma4/tt/generator.py:1332`) but no reload decision reads it, so a bucket switch under `reload_inputs=False` selects another bucket's buffers.
- Galaxy seed handling — `reset_seed_from_slots_if_needed` was unconditional (`main:1714`), now gated behind `elif reload_sampling_params` (`llama3_70b_galaxy/tt/generator.py:1707`); and `align_seed_counters_to_positions` moved from `if reset_inputs` (`main:1715-1716`) to inside `if reset_sampling_state` only, with `reload_inputs` dropped from the signature entirely (`:1639`) — tt_transformers deliberately kept it for exactly this.

Happy to send the rest of the review separately; the above are the items tied to the CI ask.

**Review by mtairum — APPROVED** [review]
Approved to unblock but please address my previous comment (+ claude suggestion):

Rebase, remove unnecessary code change on deprecated file, re-test vLLM CI to make sure that the hang previously seen is not there anymore or caused by this PR

**Review by atupe-tt — APPROVED** [review]

**Comment by djordje-tt** [comment]
LGTM! Please check vllm pipelines for sampling params verification.

**Review by djordje-tt — APPROVED** [review]

**Review by philei-tt — APPROVED** [review]

**Review by jmalone-tt — APPROVED** [review]

**Review by ssanjayTT — APPROVED** [review]

File: models/demos/t3000/llama2_70b/tt/generator_vllm.py
Comment (tchedaTT) [reply models/demos/t3000/llama2_70b/tt/generator_vllm.py:1]: Rebased onto current main at 68107a0c9b1. The removed T3000 adapter stays deleted; the PR no longer modifies or restores this path.

**Comment by tchedaTT** [comment]
@mtairum @djordje-tt Refresh complete:

- rebased onto current `main` at `68107a0c9b1`; new head is `8b263d2239f`
- the deleted T3000 adapter stays deleted
- the mode and batch concerns are covered by the merged plugin contract: sampling-mode changes are explicit planner transitions, and decode membership or batch changes set `decode_layout_changed`; either makes `reload_inputs=True` before the corresponding mode or Gemma trace buffer is replayed
- the model-side contract regression deliberately switches between existing mode and Gemma batch buckets and verifies the selected buffer is fully refreshed
- Galaxy separated sampling now carries the same `reload_inputs` authority as its forward; it only realigns seed counters from authoritative host positions, while steady async steps advance the resident counter
- the DeepSeek partial-admission and Galaxy remap review findings remain fixed after the rebase, with focused regressions

Local `pre-commit` over the full rebased diff, syntax compilation, and `git diff --check` are clean.

Fresh vLLM model validation against merged plugin `main` at `7250ddfaa988cc7417f518266dce52e425745472` is running across all models, SKUs, and tiers: https://github.com/tenstorrent/tt-metal/actions/runs/35680811729

I also updated the PR description with the new SHAs and run link.
